src/types/list.rs

```diff
@@ -750,6 +807,27 @@ impl<'py> Iterator for BoundListIterator<'py> {
             None
         })
     }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, feature = "nightly"))]
```

Comment:
```suggestion
    #[cfg(feature = "nightly")]
```

I think the critical section is necessary though, since another thread might try to use the list while you're in `advance_by`. You don't need to be in a `Py_GIL_DISABLED` cfg block to use a critical section, it's a no-op on the GIL-enabled build.